File: rater/main/rateTeams.m
function [rOptions rOutput] = rateTeams(rOptions, rOutput)
  tTree = rOutput.tTree;
  mTree = rOutput.mTree;
  mi = rOutput.mi;
  mi = rOutput.mi;
  tTree = resetRatings(tTree);
  mi = mi.reset();
  
  while (mi.hasNext())
    [mi match] = mi.next();
    [A match] = computeStrPrereqs(tTree, match, rOptions);
    [tTree match] = updateStr(tTree, match, A, rOptions);
    rOutput = updateCost(rOutput, rOptions, match, A);
    mList = mTree(match.date);
    mList(match.i) = match;
    mTree(match.date) = mList;
  end
  
  rOutput.tTree = tTree;
  rOutput.mTree = mTree;
  rOutput.mi = mi;
end

function [A match] = computeStrPrereqs(tTree, match, rOptions)
  homeTeam = tTree(match.teamNames{1});
  awayTeam = tTree(match.teamNames{2});
  fTree = rOptions.fTree;
  
  if (length(homeTeam.str) == 0)
    homeTeam.str = fTree(homeTeam.fName);
  end
  
  if (length(awayTeam.str) == 0)
    awayTeam.str = fTree(awayTeam.fName);
  end
  
  match.teamStr = [homeTeam.str; awayTeam.str];
  k = rOptions.k;
  goals = match.goalsNorm;
  A = [0 goals(2); goals(1) 0];
  t = match.days - [homeTeam.updateDays awayTeam.updateDays]';
  match.teamXP = expDecay(t, k / 365, [homeTeam.xp awayTeam.xp]');
end

function [tTree match strPost] = updateStr(tTree, match, A, rOptions)
  homeTeam = tTree(match.teamNames{1});
  awayTeam = tTree(match.teamNames{2});
  str = match.teamStr;
  strPost = computeStr(A, str, rOptions.c, ...
      rOptions.tolRel, rOptions.tolScale);
  match.teamStrPost = strPost;
  teamXP = match.teamXP;
  alphas = 1 ./ (1 + teamXP);
  strNext = computeStrNext(str, strPost, alphas);
  match.teamStrNext = strNext;
  homeTeam.str = strNext(1, :);
  awayTeam.str = strNext(2, :);
  homeTeam.xp = teamXP(1) + 1;
  awayTeam.xp = teamXP(2) + 1;
  homeTeam.updateDate = match.date;
  awayTeam.updateDate = match.date;
  homeTeam.updateDays = match.days;
  awayTeam.updateDays = match.days;
  tTree(match.teamNames{1}) = homeTeam;
  tTree(match.teamNames{2}) = awayTeam;
end

File: rater/main/rateTeams.R
rateTeams <- function(x, rOptions, rOutput, lambdas=rep(1, 4)) {
  print(x)
  strFsNorm <- x[-c(1: 6)]
  rOptions <- updateOptions(rOptions, x[c(1, 2)], x[3], x[c(4, 5)],
      x[6], strFsNorm)
  rOutput <- getRatings(rOptions, rOutput)
  strCost <- rOutput$strCost
  goalsCost <- computeGoalsCost(rOutput)
  strMeanCost <- computeStrMeanCost(rOutput)
  strFsNormCost <- norm(matrix(strFsNorm), "f")
  xpCost <- rOutput$xpCost
  goalsReg <- lambdas[1] * min(0, 1.1 - goalsCost) ^ 2
  strReg <- lambdas[2] * (min(0, 0.05 - strMeanCost[1]) ^ 2 +
      min(0, 0.05 - strMeanCost[1]) ^ 2)
  strFsNormReg <- lambdas[3] * min(0, 1 - strFsNormCost) ^ 2
  xpReg <- lambdas[4] * xpCost
  print(c(strCost, goalsReg, strReg, strFsNormReg, xpCost))
  rOutput$y <- strCost + goalsReg + strReg + strFsNormReg + xpReg
  rData <- list(rOptions=rOptions, rOutput=rOutput)
  rData
}

getRatings <- function(rOptions, rOutput) {
  tTree <- rOutput$tTree
  gTree <- rOutput$gTree
  gi <- rOutput$gi
  tTree <- resetRatings(tTree)
  gi <- reset(gi)
  gamePrev <- NULL
  i <- 1
  
  while (hasNextGame(gi)) {
    gameData <- nextGame(gi)
    gi <- gameData[["gi"]]
    game <- gameData[["game"]]
    strPrereqs <- computeStrPrereqs(tTree, game, rOptions)
    updateStrData <- updateStr(strPrereqs, rOptions)
    tTree <- updateStrData[["tTree"]]
    game <- updateStrData[["game"]]
    costData <- updateCost(rOptions, rOutput, game, gamePrev)
    rOutput <- costData[["rOutput"]]
    game <- costData[["game"]]
    gDateList <- gTree[[game$gameDateStr]]
    gDateList[[game$gameNum]] <- game
    gTree[game$gameDateStr] <- gDateList
    gamePrev <- game
    i <- i + 1
  }
  
  rOutput$tTree <- tTree
  rOutput$gTree <- gTree
  rOutput$gi <- gi
  rOutput
}

resetRatings <- function(tTree) {
  teams <- keys(tTree)
  n <- length(teams)
  i <- 1
  
  while (i <= n) {
    teamName <- teams[i]
    team <- tTree[[teamName]]
    tTree[teamName] <- resetTeam(team)
    i <- i + 1
  }
  
  tTree
} 

computeStrPrereqs <- function(tTree, game, rOptions) {
  homeTeamName <- game$teamNames[1]
  awayTeamName <- game$teamNames[2]
  homeTeam <- tTree[[homeTeamName]]
  awayTeam <- tTree[[awayTeamName]]
  fTree <- rOptions$fTree 
  game <- updateGamePreRate(game, fTree, rOptions$ks,
      homeTeam, awayTeam)
  strPrereqs <- list(game=game, tTree=tTree)
  strPrereqs
}

updateStr <- function(strPrereqs, rOptions) {
  game <- strPrereqs[["game"]]
  tTree <- strPrereqs[["tTree"]]
  strPost <- computeStr(game$A, game$teamStr, rOptions$c,
      rOptions$tolRel, rOptions$tolScale)
  game <- updateGamePostRate(game, strPost)
  homeTeamName <- game$teamNames[1]
  awayTeamName <- game$teamNames[2]
  homeTeam <- tTree[[homeTeamName]]
  awayTeam <- tTree[[awayTeamName]] 
  tTree[homeTeamName] <- updateTeam(homeTeam, game, 1)
  tTree[awayTeamName] <- updateTeam(awayTeam, game, 2)
  updateStrData <- list(tTree=tTree, game=game)
  updateStrData
}
